src/frontend/js/components/PaymentButton/index.tsx

```diff
@@ -0,0 +1,226 @@
+import { useEffect, useMemo, useRef, useState } from 'react';
+import { defineMessages, FormattedMessage, useIntl } from 'react-intl';
+import type * as Joanie from 'types/Joanie';
+import { OrderState } from 'types/Joanie';
+import type { Nullable } from 'types/utils';
+import { useCourseCode } from 'data/CourseCodeProvider';
+import { Spinner } from 'components/Spinner';
+import PaymentInterface from 'components/PaymentInterfaces';
+import { useOrders } from 'hooks/useOrders';
+import { useJoanieApi } from 'data/JoanieApiProvider';
+
+const messages = defineMessages({
+  pay: {
+    defaultMessage: 'Pay {price}',
+    description: 'CTA label to proceed of the product',
+    id: 'components.PaymentButton.pay',
+  },
+  payInOneClick: {
+    defaultMessage: 'Pay in one click {price}',
+    description: 'CTA label to proceed of the product',
+    id: 'components.PaymentButton.payInOneClick',
+  },
+  errorDefault: {
+    defaultMessage: 'An error occurred during payment. Please retry later.',
+    description: 'Error message shown when payment creation request failed.',
+    id: 'components.PaymentButton.errorDefault',
+  },
+  errorAbort: {
+    defaultMessage: 'You have aborted the payment.',
+    description: 'Error message shown when user aborts the payment.',
+    id: 'components.PaymentButton.errorAbort',
+  },
+  errorAborting: {
+    defaultMessage: 'Aborting the payment...',
+    description: 'Error message shown when user asks to abort the payment.',
+    id: 'components.PaymentButton.errorAborting',
+  },
+  paymentInProgress: {
+    defaultMessage: 'Payment in progress',
+    description: 'Label for screen reader when a payment is in progress.',
+    id: 'components.PaymentButton.paymentInProgress',
+  },
+});
+
+export enum PaymentErrorMessageId {
+  ERROR_DEFAULT = 'errorDefault',
+  ERROR_ABORT = 'errorAbort',
+  ERROR_ABORTING = 'errorAborting',
+}
+
+interface PaymentButtonProps {
+  product: Joanie.Product;
+  billingAddress?: Joanie.Address;
+  creditCard?: Nullable<Joanie.CreditCard['id']>;
+  onSuccess: () => void;
+}
+
+type PaymentInfo = Joanie.Payment & { order_id: string };
+type OneClickPaymentInfo = Joanie.PaymentOneClick & { order_id: string };
+
+enum ComponentStates {
+  IDLE = 'idle',
+  LOADING = 'loading',
+  ERROR = 'error',
+}
```

Comment:
shouldn't we use a file outside like type/Joanie.ts or a one specific for the payment?

Reply:
These states are related to the payment button that's why I decided to let this declaration here.